File: AzureBotService/dialogs/mainDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const https = require('https');
const { CardFactory, AttachmentLayoutTypes } = require('botbuilder');
const { ConfirmPrompt, DialogSet, DialogTurnStatus, OAuthPrompt, TextPrompt, WaterfallDialog } = require('botbuilder-dialogs');

const { LogoutDialog } = require('./logoutDialog');
const { OAuthSAMLHelper } = require('../oauthSAMLHelper');

const CONFIRM_PROMPT = 'ConfirmPrompt';
const MAIN_DIALOG = 'MainDialog';
const MAIN_WATERFALL_DIALOG = 'MainWaterfallDialog';
const OAUTH_PROMPT = 'OAuthPrompt';
const PRODUCT_PROMPT = 'ProductPrompt';

class MainDialog extends LogoutDialog {
    constructor() {
        super(MAIN_DIALOG, process.env.connectionName);

        this.addDialog(new OAuthPrompt(OAUTH_PROMPT, {
            connectionName: process.env.connectionName,
            text: 'Please Sign In',
            title: 'Sign In',
            timeout: 300000
        }));
        this.addDialog(new TextPrompt(PRODUCT_PROMPT));
        this.addDialog(new ConfirmPrompt(CONFIRM_PROMPT));
        this.addDialog(new WaterfallDialog(MAIN_WATERFALL_DIALOG, [
            this.promptStep.bind(this),
            this.loginStep.bind(this),
            this.displayTokenPhase1.bind(this),
            this.displayTokenPhase2.bind(this),
            this.productSearch.bind(this)
        ]));

        this.initialDialogId = MAIN_WATERFALL_DIALOG;
    }

    /**
     * The run method handles the incoming activity (in the form of a DialogContext) and passes it through the dialog system.
     * If no dialog is active, it will start the default dialog.
     * @param {*} dialogContext
     */
    async run(context, accessor) {
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this);

        const dialogContext = await dialogSet.createContext(context);
        const results = await dialogContext.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            await dialogContext.beginDialog(this.id);
        }
    }

    async promptStep(stepContext) {
        return await stepContext.beginDialog(OAUTH_PROMPT);
    }

    async loginStep(stepContext) {
        // Get the token from the previous step. Note that we could also have gotten the
        // token directly from the prompt itself. There is an example of this in the next method.
        const tokenResponse = stepContext.result;
        if (tokenResponse) {
            return await stepContext.prompt(CONFIRM_PROMPT, 'Would you like to view your token?');
        }
        await stepContext.context.sendActivity('Login was not successful please try again.');
        return await stepContext.endDialog();
    }

    async displayTokenPhase1(stepContext) {
        await stepContext.context.sendActivity('Thank you.');

        const result = stepContext.result;
        if (result) {
            // Call the prompt again because we need the token. The reasons for this are:
            // 1. If the user is already logged in we do not need to store the token locally in the bot and worry
            // about refreshing it. We can always just call the prompt again to get the token.
            // 2. We never know how long it will take a user to respond. By the time the
            // user responds the token may have expired. The user would then be prompted to login again.
            //
            // There is no reason to store the token locally in the bot because we can always just call
            // the OAuth prompt to get the token or get a new token if needed.
            return await stepContext.beginDialog(OAUTH_PROMPT);
        }
        return await stepContext.endDialog();
    }

    async displayTokenPhase2(stepContext) {
        const aadTokenResponse = stepContext.result;
        if (aadTokenResponse) {
            await stepContext.context.sendActivity(`Here is your OAuth access token ${ aadTokenResponse.token }`);
            stepContext.values.token = aadTokenResponse;
            return await stepContext.prompt(PRODUCT_PROMPT, { prompt: 'Which product are you looking for?' }); 
        }
        return await stepContext.endDialog();
    }

    async productSearch(stepContext) {
        if (stepContext.result) {            
            var products = null;
            const aadTokenResponse = stepContext.values.token;
            if (aadTokenResponse && aadTokenResponse.token) {
                var oauthSAMLHelper = new OAuthSAMLHelper();
                var xsuaaAccessToken = await oauthSAMLHelper.getSAMLAssertionFromAAD(aadTokenResponse.token).then(oauthSAMLHelper.getAccessTokenFromXSUAA);

                var productResultSet = new Promise(function(resolve, reject) {
                    console.log("Now calling iFlow with access token " + xsuaaAccessToken);
                    https.get({
                        protocol: 'https:',
                        hostname: process.env.cpiHost,
                        port: 443,                       
                        path: process.env.cpiPath,
                        headers: {                   
                            'Authorization': 'Bearer ' + xsuaaAccessToken,
                            'productId' : stepContext.result
                        }
                    },                    
                    (response) => {
                        let body = '';
                        response.on('data', function(data) {
                            body += String(data);
                        });
    
                        response.on('end', function() {
                            // console.log(body);
                            // Success finally return the resultset
                            resolve(JSON.parse(body));
                        });
                    }).on('error', (err) => {
                        console.log('Error: ' + err.message);
                        reject(new Error('Error: ' + err.message));
                    });
                });                
                
                products = await productResultSet;                
                const numberOfProducts = products.d.results.length;

                // Get the images from the public SAP E5 system
                var imageURL = 'https://sapes5.sapdevcenter.com';

                // Create a hero card and loop over products result set
                // https://docs.microsoft.com/en-us/adaptive-cards/
                const reply = { attachments: [], attachmentLayout: AttachmentLayoutTypes.List };
                for (let cnt = 0; cnt < numberOfProducts; cnt++) {
                    var product = products.d.results[cnt];
                    var productImageURL = imageURL + product.ProductPictureURL;
                    const card = CardFactory.heroCard(
                        'Product Name: ' + product.Product,
                        'Category: ' + product.ProductCategory,
                        [{ type: 'Image', alt: 'SAP Logo', url: productImageURL, height: '5px', width: '5px' }],
                        ['Order via email'],
                        { subtitle: `Price : ${ product.Price } ${ product.Currency }` }
                    );
                    reply.attachments.push(card);
                }                
                await stepContext.context.sendActivity(reply);                
            } else {
                await stepContext.context.sendActivity('We couldn\'t log you in. Please try again later.');
            }
        }
        return await stepContext.endDialog();   
    }    
}

module.exports.MainDialog = MainDialog;

File: ProductSearchBot/Dialogs/MainDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Net.Http;
using System.Net.Http.Headers;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

namespace Microsoft.BotBuilderSamples
{
    
    public class MainDialog : LogoutDialog
    {
        private readonly ILogger _logger;
        private SCPClient scpClient;
        private readonly string _cpiHost;
        private readonly string _cpiPath;

        public MainDialog(IConfiguration configuration, ILogger<MainDialog> logger)
            : base(nameof(MainDialog), configuration["ConnectionName"])
        {
            _logger = logger;

            scpClient = new SCPClient(configuration);

            _cpiHost = configuration["CPIBaseURI"];
            _cpiPath = configuration["CPIPath"];

            AddDialog(new TokenExchangeOAuthPrompt(
                nameof(TokenExchangeOAuthPrompt),
                new OAuthPromptSettings
                {
                    ConnectionName = ConnectionName,
                    Text = "Please Sign In",
                    Title = "Sign In",
                    Timeout = 1000 * 60 * 1 // User has 5 minutes to login (1000 * 60 * 5)
                    //EndOnInvalidMessage = true
                }));

            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));

            AddDialog(new TextPrompt(nameof(TextPrompt)));

            AddDialog(new WaterfallDialog(nameof(WaterfallDialog),  new WaterfallStep[] { PromptStepAsync, LoginStepAsync, DisplayTokenAsync, AskForProductAsync, DisplaySearchResultsAsync }));

            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> PromptStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            return await stepContext.BeginDialogAsync(nameof(TokenExchangeOAuthPrompt), null, cancellationToken);
        }

        private async Task<DialogTurnResult> LoginStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // Get the token from the previous step. Note that we could also have gotten the
            // token directly from the prompt itself. There is an example of this in the next method.
            var tokenResponse = (TokenResponse)stepContext.Result;
            if (tokenResponse?.Token != null)
            {
                // store token for later use
                stepContext.Values.Add("aadToken", tokenResponse.Token);
                await stepContext.Context.SendActivityAsync($"You're logged in");
                return await stepContext.PromptAsync(nameof(ConfirmPrompt), new PromptOptions { Prompt = MessageFactory.Text("Would you like to view your token?") }, cancellationToken);
            }
            await stepContext.Context.SendActivityAsync(MessageFactory.Text("Login was not successful please try again."), cancellationToken);
            return await stepContext.EndDialogAsync(cancellationToken: cancellationToken);
        }

        private async Task<DialogTurnResult> DisplayTokenAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {   
            var result = (bool)stepContext.Result;
            if (result) {
                await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Here is your token {stepContext.Values["aadToken"].ToString()}"), cancellationToken);
            }
            return await stepContext.ContinueDialogAsync(cancellationToken: cancellationToken);
        }

        private async Task<DialogTurnResult> AskForProductAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            return await stepContext.PromptAsync(nameof(TextPrompt), new PromptOptions{ Prompt = MessageFactory.Text("Which product are you looking for?") }, cancellationToken);
        }

        private async Task<DialogTurnResult> DisplaySearchResultsAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (stepContext.Result != null) {  
                _logger.LogInformation("Search started for product name: " + stepContext.Result);
                string aadAccessToken = stepContext.Values["aadToken"].ToString();
                _logger.LogInformation("Using AAD OAuth token: " + aadAccessToken);
                var samlAssertion = await scpClient.GetSAMLAssertionFromAAD(aadAccessToken);
                _logger.LogInformation("Received SAML assertion from AAD: " + samlAssertion);
                var scpAccessToken = await scpClient.GetAccessTokenFromSCP(samlAssertion);
                _logger.LogInformation("Received access token from SCP: " + scpAccessToken);
                
                using (HttpClient httpClient = new HttpClient())
                {
                    httpClient.BaseAddress = new Uri(_cpiHost);
                    httpClient.DefaultRequestHeaders.Accept.Clear();
                    httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", scpAccessToken);
                    httpClient.DefaultRequestHeaders.Add("productId", stepContext.Result.ToString());
                    var httpResponse = await httpClient.GetAsync(_cpiPath);

                    httpResponse.EnsureSuccessStatusCode(); // throws if not 200-299

                    if (httpResponse.Content is object && httpResponse.Content.Headers.ContentType.MediaType == "application/json")
                    {
                        var responseBody = httpResponse.Content.ReadAsStringAsync().Result;
                        try
                        {
                            var responseJSON = JObject.Parse(responseBody);
		                    var products = responseJSON.SelectTokens("d.results[*]");
                            var imageURL = "https://sapes5.sapdevcenter.com";
                            var attachments = new List<Attachment>();
                            foreach (var product in products)
                            {
                                var title = String.Format("{0} ({1} {2})", product["Product"], product["Price"], product["Currency"]);
                                var subtitle = product["ProductCategory"].ToString();
                                var imageUrl = imageURL + product["ProductPictureURL"];

                                CardAction orderButton = new CardAction()
                                {
                                    Value = "https://www.microsoft.com/",
                                    Type = ActionTypes.OpenUrl,
                                    Title = "Order now"
                                };

                                var heroCard = new HeroCard
                                {  
                                    Title = title,
                                    Subtitle = subtitle,
                                    Images = new List<CardImage> { new CardImage(imageUrl) },
                                    Buttons =  new List<CardAction> { orderButton }
                                };
                                attachments.Add(heroCard.ToAttachment());
                            }

                            if (attachments.Count > 0) {
                                var activity = MessageFactory.Attachment(attachments);
                                await stepContext.Context.SendActivityAsync(activity, cancellationToken);
                            } else {
                                await stepContext.Context.SendActivityAsync(MessageFactory.Text("Sorry, no products found!"), cancellationToken);
                            }
                        }                                 
                        catch (Newtonsoft.Json.JsonException) // Invalid JSON
                        {
                            throw new InvalidOperationException("No JSON response. Access Token request failed");
                        }  
                    }
                    else
                    {
                        throw new InvalidOperationException("HTTP Response was invalid and cannot be deserialized.");
                    }
                }
            }
            return await stepContext.EndDialogAsync(cancellationToken: cancellationToken);
        }
    }
}
